hci.go

```diff
@@ -188,33 +192,65 @@ func (h *hci) poll() error {
 	h.transport.startRead()
 	defer h.transport.endRead()
 
-	i := 0
-	for h.transport.Buffered() > 0 {
-		sz := h.transport.Buffered()
-		c := sz + 4 - (sz % 4)
-		_, err := h.transport.Read(h.buf[i : i+c])
-		if err != nil {
-			return err
+	for h.transport.Buffered() > 0 || h.end > h.pos {
+		// perform read only if more data is available
+		available := h.transport.Buffered()
+		if available > 0 {
+			aligned := available + (4-(available%4))%4
+			n, err := h.transport.Read(h.buf[h.end : h.end+aligned])
+			if err != nil {
+				return err
+			}
+			h.end += n
 		}
-		i += sz
 
-		done, err := h.processPacket(i)
+		// do processing
+		h.pos += 1
+		done, err := h.processPacket()
 		switch {
-		case err == ErrHCIUnknown || err == ErrHCIInvalidPacket || err == ErrHCIUnknownEvent:
+		case err == ErrHCIInvalidPacket:
+			if debug {
+				println("hci poll invalid packet:", err.Error(), hex.EncodeToString(h.buf[:h.pos]))
+			}
+
+			time.Sleep(5 * time.Millisecond)
+		case err == ErrHCIUnknown || err == ErrHCIUnknownEvent:
 			if debug {
-				println("hci error:", err.Error(), hex.EncodeToString(h.buf[:i]))
+				println("hci poll unknown packet:", err.Error(), hex.EncodeToString(h.buf[:h.pos]))
 			}
-			i = 0
+
+			h.pos = 0
+			h.end = 0
+
 			time.Sleep(5 * time.Millisecond)
 		case err != nil:
+			// some other error, so return
+			h.pos = 0
+			h.end = 0
+
 			return err
 		case done:
+			// move remaining data to start of buffer
```

Comment:
Could some ring buffer implementation be reused here instead? It looks like `machine` has one https://github.com/tinygo-org/tinygo/blob/2a76ceb7dd5ea5a834ec470b724882564d9681b3/src/machine/buffer.go#L9.

Reply:
An interesting idea, but I think out of scope for this specific PR.